db/db_impl/db_impl_write.cc

```diff
@@ -1737,6 +1737,178 @@ void DBImpl::NotifyOnMemTableSealed(ColumnFamilyData* /*cfd*/,
+  // If there is valid data in the memtable,
+  // or at least range tombstones, copy over the info
+  // to the new memtable.
+  if (iter->Valid() || !range_del_agg->IsEmpty()) {
+    std::unique_ptr<CompactionFilter> compaction_filter;
+    if (ioptions->compaction_filter_factory != nullptr &&
+        ioptions->compaction_filter_factory->ShouldFilterTableFileCreation(
+            TableFileCreationReason::kFlush)) {
+      CompactionFilter::Context ctx;
+      ctx.is_full_compaction = false;
+      ctx.is_manual_compaction = false;
+      ctx.column_family_id = cfd->GetID();
+      ctx.reason = TableFileCreationReason::kFlush;
+      compaction_filter =
+          ioptions->compaction_filter_factory->CreateCompactionFilter(ctx);
+      if (compaction_filter != nullptr &&
+          !compaction_filter->IgnoreSnapshots()) {
+        s = Status::NotSupported(
+            "CompactionFilter::IgnoreSnapshots() = false is not supported "
+            "anymore.");
+        return s;
+      }
+    }
+
+    Env* env = immutable_db_options_.env;
+    assert(env);
+    MergeHelper merge(
+        env, (cfd->internal_comparator()).user_comparator(),
+        (ioptions->merge_operator).get(), compaction_filter.get(),
+        ioptions->logger, true /* internal key corruption is not ok */,
+        snapshot_seqs.empty() ? 0 : snapshot_seqs.back(), snapshot_checker);
+    CompactionIterator c_iter(
+        iter.get(), (cfd->internal_comparator()).user_comparator(), &merge,
+        kMaxSequenceNumber, &snapshot_seqs, earliest_write_conflict_snapshot,
+        snapshot_checker, env, ShouldReportDetailedTime(env, ioptions->stats),
+        true /* internal key corruption is not ok */, range_del_agg.get(),
+        nullptr, ioptions->allow_data_in_errors,
+        /*compaction=*/nullptr, compaction_filter.get(),
+        /*shutting_down=*/nullptr,
+        /*preserve_deletes_seqnum=*/0, /*manual_compaction_paused=*/nullptr,
+        /*manual_compaction_canceled=*/nullptr, immutable_db_options_.info_log,
+        &(cfd->GetFullHistoryTsLow()));
+
+    c_iter.SeekToFirst();
+
+    mutex_.AssertHeld();
+
+    // Set earliest sequence number in the new memtable
+    // to be equal to the earliest sequence number of the
+    // memtable being flushed (See later if there is a need
+    // to update this number!).
+    new_mem->SetEarliestSequenceNumber(earliest_seqno);
+    SequenceNumber new_earliest_seqno = kMaxSequenceNumber;
+    SequenceNumber new_first_seqno = kMaxSequenceNumber;
+
+    // Key transfer
+    for (; c_iter.Valid(); c_iter.Next()) {
+      const ParsedInternalKey ikey = c_iter.ikey();
+      const Slice value = c_iter.value();
+      new_earliest_seqno = ikey.sequence < new_earliest_seqno
+                               ? ikey.sequence
+                               : new_earliest_seqno;
+      new_first_seqno =
+          ikey.sequence < new_first_seqno ? ikey.sequence : new_first_seqno;
+      new_mem->SetFirstSequenceNumber(new_first_seqno);
```

Comment:
Can we track ```new_first_seqno``` and call ```SetFirstSequenceNumber``` in the end since its an atomic store?

Reply:
Good call, I just made the edit - thanks for the catch